src/netops.c

```diff
@@ -206,83 +206,100 @@ void gitno_connection_data_free_ptrs(gitno_connection_data *d)
 	git__free(d->pass); d->pass = NULL;
 }
 
-#define hex2c(c) ((c | 32) % 39 - 9)
-static char* unescape(char *str)
-{
-	int x, y;
-	int len = (int)strlen(str);
-
-	for (x=y=0; str[y]; ++x, ++y) {
-		if ((str[x] = str[y]) == '%') {
-			if (y < len-2 && isxdigit(str[y+1]) && isxdigit(str[y+2])) {
-				str[x] = (hex2c(str[y+1]) << 4) + hex2c(str[y+2]);
-				y += 2;
-			}
-		}
-	}
-	str[x] = '\0';
-	return str;
-}
-
 int gitno_extract_url_parts(
-		char **host,
-		char **port,
-		char **path,
-		char **username,
-		char **password,
-		const char *url,
-		const char *default_port)
+	char **host_out,
+	char **port_out,
+	char **path_out,
+	char **username_out,
+	char **password_out,
+	const char *url,
+	const char *default_port)
 {
 	struct http_parser_url u = {0};
-	const char *_host, *_port, *_path, *_userinfo;
+	bool has_host, has_port, has_path, has_userinfo;
+	const char *url_host, *url_port, *url_path, *url_userinfo;
+	size_t url_host_len, url_port_len, url_path_len, url_userinfo_len;
+	git_buf host = GIT_BUF_INIT,
+		port = GIT_BUF_INIT,
+		path = GIT_BUF_INIT,
+		username = GIT_BUF_INIT,
+		password = GIT_BUF_INIT;
+	int error = 0;
 
 	if (http_parser_parse_url(url, strlen(url), false, &u)) {
 		giterr_set(GITERR_NET, "malformed URL '%s'", url);
-		return GIT_EINVALIDSPEC;
+		error = GIT_EINVALIDSPEC;
+		goto done;
 	}
 
-	_host = url+u.field_data[UF_HOST].off;
-	_port = url+u.field_data[UF_PORT].off;
-	_path = url+u.field_data[UF_PATH].off;
-	_userinfo = url+u.field_data[UF_USERINFO].off;
+	has_host = !!(u.field_set & (1 << UF_HOST));
+	url_host = url + u.field_data[UF_HOST].off;
+	url_host_len = u.field_data[UF_HOST].len;
 
-	if (u.field_set & (1 << UF_HOST)) {
-		*host = git__substrdup(_host, u.field_data[UF_HOST].len);
-		GITERR_CHECK_ALLOC(*host);
-	}
+	has_port = !!(u.field_set & (1 << UF_PORT));
+	url_port = url + u.field_data[UF_PORT].off;
+	url_port_len = u.field_data[UF_PORT].len;
 
-	if (u.field_set & (1 << UF_PORT))
-		*port = git__substrdup(_port, u.field_data[UF_PORT].len);
-	else
-		*port = git__strdup(default_port);
-	GITERR_CHECK_ALLOC(*port);
+	has_path = !!(u.field_set & (1 << UF_PATH));
+	url_path = url + u.field_data[UF_PATH].off;
+	url_path_len = u.field_data[UF_PATH].len;
 
-	if (path) {
-		if (u.field_set & (1 << UF_PATH)) {
-			*path = git__substrdup(_path, u.field_data[UF_PATH].len);
-			GITERR_CHECK_ALLOC(*path);
-		} else {
-			git__free(*port);
-			*port = NULL;
-			git__free(*host);
-			*host = NULL;
-			giterr_set(GITERR_NET, "invalid url, missing path");
-			return GIT_EINVALIDSPEC;
-		}
+	has_userinfo = !!(u.field_set & (1 << UF_USERINFO));
+	url_userinfo = url + u.field_data[UF_USERINFO].off;
+	url_userinfo_len = u.field_data[UF_USERINFO].len;
+
+	if (has_host)
```

Comment:
Dunno. Wouldn't this function be easier to read if setting the buffers would be all in one spot? E.g.

```
if (!!(u.field_set & (1 << UF_HOST)))
        git_buf_decode_percent(&host, url + u.field_data[UF_HOST].off, u.field_data[UF_HOST].len);
```

A lot less variables and all handling of that information is at exactly one spot.

Reply:
The lines are soooooo long then.  It is much cleaner to when we use some temporary variables.

Reply:
Certainly true. On the other hand, you could just have locale variables inside of that block